File: problems/first-bad-version.js
function isBadVersion(n) {}

function firstBadVersion(n) {
  let start = 1;
  let end = n;
  let bad;

  while (start <= end) {
    const mid = Math.floor((start + end) / 2);
    if (isBadVersion(mid)) {
      bad = mid;
      end = mid - 1;
    } else start = mid + 1;
  }

  return bad;
}

File: medium/first-bad-version.py
# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:

class Solution:
    def firstBadVersion(self, n: int) -> int:
        s = 1
        e = n

        while s <= e:
            mid = (s + e) // 2
            
            if not isBadVersion(mid): s = mid + 1
            else: e = mid - 1
        
        return s